cpp/src/arrow/python/inference.cc

```diff
@@ -354,12 +365,15 @@ class TypeInferrer {
       *keep_going = make_unions_;
     } else if (PyArray_CheckAnyScalarExact(obj)) {
       RETURN_NOT_OK(VisitDType(PyArray_DescrFromScalar(obj), keep_going));
-    } else if (PyList_Check(obj)) {
-      RETURN_NOT_OK(VisitList(obj, keep_going));
+    } else if (PySet_Check(obj) || (Py_TYPE(obj) == &PyDictValues_Type)) {
+      RETURN_NOT_OK(VisitSet(obj, keep_going));
     } else if (PyArray_Check(obj)) {
       RETURN_NOT_OK(VisitNdarray(obj, keep_going));
     } else if (PyDict_Check(obj)) {
       RETURN_NOT_OK(VisitDict(obj));
+    } else if (PyList_Check(obj) || PyTuple_Check(obj) ||
+               PyObject_IsInstance(obj, deque_type_.obj())) {
```

Comment:
I still don't understand why we want to accept deques here. I don't think it is usual to store analytical data in deques. @jorisvandenbossche What do you think?

Reply:
I personally agree it doesn't seem needed to add special case code in C++ to handle deques. 

I would actually even doubt whether we should handle sets (they are unordered, so not necessarily directly mapping to an array, the order you get from iteration / conversion to list is not deterministic/guaranteed AFAIK). But since we already support sets in non-nested situations, I understand that this is making that consistent.

Reply:
I supported deques just because we already support them in non nested situation. My goal was to keep the same support for nested and non-nested cases. Otherwise it seemed odd that we support `deque([1, 2, 3])` but not `[deque([1, 2, 3])]`

Reply:
I think the idea is that we support generic iterables as top-level input, so that you can for example call `pa.array` on a generator if you want to generate data lazily. It makes little sense to allow the same flexibility for inner values, though, such as values inside a list array.
```python
>>> pa.array((x for x in (1, 2, 3)))
<pyarrow.lib.Int64Array object at 0x7f44678606e0>
[
  1,
  2,
  3
]
```

Reply:
Removed support for nested `deque`